translator/opencensus/oc_to_internal.go

```diff
@@ -0,0 +1,388 @@
+			conventions.OcTimeEventMessageEventUSize: attrs[2],
+			conventions.OcTimeEventMessageEventCSize: attrs[3],
+		},
+		0)
+}
+
+func truncableStringToStr(ts *octrace.TruncatableString) string {
+	if ts == nil {
+		return ""
+	}
+	return ts.Value
+}
+
+func ocNodeResourceToInternal(ocNode *occommon.Node, ocResource *ocresource.Resource) *data.Resource {
+	resource := &data.Resource{}
+
+	// Number of special fields in the Node. See the code below that deals with special fields.
+	const specialNodeAttrCount = 7
+
+	// Number of special fields in the Resource.
+	const specialResourceAttrCount = 1
+
+	// Calculate maximum total number of attributes. It is OK if we are a bit higher than
+	// the exact number since this is only needed for capacity reservation.
+	maxTotalAttrCount := 0
+	if ocNode != nil {
+		maxTotalAttrCount += len(ocNode.Attributes) + specialNodeAttrCount
+	}
+	if ocResource != nil {
+		maxTotalAttrCount += len(ocResource.Labels) + specialResourceAttrCount
+	}
+
+	// Create a map where we will place all attributes from the Node and Resource.
+	attrs := make(data.AttributesMap, maxTotalAttrCount)
+
+	if ocNode != nil {
+		// Copy all Attributes.
+		for k, v := range ocNode.Attributes {
+			attrs[k] = data.NewAttributeValueString(v)
+		}
+
+		// Add all special fields.
+		if ocNode.ServiceInfo != nil {
+			if ocNode.ServiceInfo.Name != "" {
+				attrs[conventions.AttributeServiceName] = data.NewAttributeValueString(
+					ocNode.ServiceInfo.Name)
+			}
+		}
+		if ocNode.Identifier != nil {
+			if ocNode.Identifier.StartTimestamp != nil {
+				attrs[conventions.OcAttributeProcessStartTime] = data.NewAttributeValueString(
+					ptypes.TimestampString(ocNode.Identifier.StartTimestamp))
+			}
+			if ocNode.Identifier.HostName != "" {
+				attrs[conventions.AttributeHostHostname] = data.NewAttributeValueString(
+					ocNode.Identifier.HostName)
+			}
+			if ocNode.Identifier.Pid != 0 {
+				attrs[conventions.OcAttributeProcessID] = data.NewAttributeValueInt(int64(ocNode.Identifier.Pid))
+			}
+		}
+		if ocNode.LibraryInfo != nil {
+			if ocNode.LibraryInfo.CoreLibraryVersion != "" {
+				attrs[conventions.AttributeLibraryVersion] = data.NewAttributeValueString(
+					ocNode.LibraryInfo.CoreLibraryVersion)
+			}
+			if ocNode.LibraryInfo.ExporterVersion != "" {
+				attrs[conventions.OcAttributeExporterVersion] = data.NewAttributeValueString(
+					ocNode.LibraryInfo.ExporterVersion)
+			}
+			if ocNode.LibraryInfo.Language != occommon.LibraryInfo_LANGUAGE_UNSPECIFIED {
+				attrs[conventions.AttributeLibraryLanguage] = data.NewAttributeValueString(
+					ocNode.LibraryInfo.Language.String())
+			}
+		}
+	}
+
+	if ocResource != nil {
+		// Copy resource Labels.
+		for k, v := range ocResource.Labels {
+			attrs[k] = data.NewAttributeValueString(v)
+		}
+		// Add special fields.
+		if ocResource.Type != "" {
+			attrs[conventions.OcAttributeResourceType] = data.NewAttributeValueString(ocResource.Type)
+		}
+	}
+
+	if len(attrs) != 0 {
```

Comment:
Not sure what you mean. This avoids setting an empty map and is a minor optimization.

Reply:
See https://github.com/open-telemetry/opentelemetry-collector/pull/596#discussion_r389985941

Reply:
Actually this check is still useful, since even if node and resource are present, they may not produce any atttributes, e.g. when they are just empty structs.

Reply:
Correct, that's exactly the purpose of the check. We don't want a map if there are no attributes.